ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/TransactionContractKeysValidation.scala

```diff
@@ -0,0 +1,295 @@
+      transactionCommitter: TransactionCommitter
+  ): Step[DamlTransactionEntrySummary] =
+    (commitContext, transactionEntry) => {
+      val damlState = commitContext
+        .collectInputs[(DamlStateKey, DamlStateValue), Map[DamlStateKey, DamlStateValue]] {
+          case (key, Some(value)) if key.hasContractKey => key -> value
+        }
+      val contractKeyDamlStateKeysToContractIds: Map[DamlStateKey, RawContractId] =
+        damlState.collect {
+          case (k, v) if k.hasContractKey && v.getContractKeyState.getContractId.nonEmpty =>
+            k -> v.getContractKeyState.getContractId
+        }
+      val contractKeyDamlStateKeys: Set[DamlStateKey] =
+        contractKeyDamlStateKeysToContractIds.keySet
+      val contractKeysToContractIds: Map[DamlContractKey, RawContractId] =
+        contractKeyDamlStateKeysToContractIds.map(m => m._1.getContractKey -> m._2)
+
+      validateTransactionContractKeysUniquenessAndConsistency(
+        transactionCommitter,
+        commitContext.recordTime,
+        transactionEntry,
+        contractKeyDamlStateKeys,
+        contractKeysToContractIds,
+      ) match {
+        case StepContinue(transactionEntry) =>
+          validateTransactionContractKeysCausalMonotonicity(
+            transactionCommitter,
+            commitContext.recordTime,
+            transactionEntry,
+            contractKeyDamlStateKeys,
+            damlState,
+          )
+        case err => err
+      }
+    }
+
+  /** LookupByKey nodes themselves don't actually fetch the contract.
+    * Therefore we need to do an additional check on all contract keys
+    * to ensure the referred contract satisfies the causal monotonicity invariant.
+    * This could be reduced to only validate this for keys referred to by
+    * NodeLookupByKey.
+    */
+  private def validateTransactionContractKeysCausalMonotonicity(
+      transactionCommitter: TransactionCommitter,
+      recordTime: Option[Timestamp],
+      transactionEntry: DamlTransactionEntrySummary,
+      keys: Set[DamlStateKey],
+      damlState: Map[DamlStateKey, DamlStateValue],
+  ): StepResult[DamlTransactionEntrySummary] = {
+    val causalKeyMonotonicity = keys.forall { key =>
+      val state = damlState(key)
+      val keyActiveAt =
+        Conversions.parseTimestamp(state.getContractKeyState.getActiveAt).toInstant
+      !keyActiveAt.isAfter(transactionEntry.ledgerEffectiveTime.toInstant)
+    }
+    if (causalKeyMonotonicity)
+      StepContinue(transactionEntry)
+    else
+      transactionCommitter.reject(
+        recordTime,
+        transactionCommitter.buildRejectionLogEntry(
+          transactionEntry,
+          RejectionReason.Inconsistent("Causal monotonicity violated"),
+        ),
+      )
+  }
+
+  private sealed abstract class KeyValidationError()
+  private case object Duplicate extends KeyValidationError
+  private case object Inconsistent extends KeyValidationError
+
+  private case class KeyValidationState(
```

Comment:
If you were to make this class monoidal, where you could merge two of them together rather than always add to an existing one, it would be much easier to separate the distinct validation checks.

Reply:
IMO forcing the monotonicity check (as any potential future keys check, if any) into the fold is only going to obfuscate it. I propose to have 2 (for now, and monadic) steps: monotonicity check and traversal-bound (aggregated) check. The traversal-bound check, in turn, has 2 (for now, and monadic) sub-steps: the uniqueness and consistency checks.

This is already the current situation; I am also going to turn the traversal-bound aggregated check state into a rather opaque monoid and that indeed makes the code (a bit) shorter and more pleasant to the eye, but that would be all. Would that be OK for you? Or what did you have in mind?

Reply:
I think that sounds like a good compromise to me!

Reply:
Done, and also the split into smaller classes.